### ISSUE TITLE
Keras / Tensorflow cumstom cross entropy loss with dynamic masking

### ISSUE THREAD

**Issue by GMarzinotto** [keras-team/keras#11248]
I'm trying to create a custom loss in Keras, using Tensorflow Backend.

This custom loss is nothing but a cross entropy loss, where some output classes are masked.

For information, the model is a recurrent neural network Sequence-to-Sequence model, but the loss should work even if it was a simple classifier. 

----------

This mask must be dynamic. 

I mean, a different mask is applied depending on the value of y_true for each data sample.

----------

Now I explain why I want to do this so I can detail masking process 

I have some label hierarchy in my problem. Lets say my labels are:

    labels = [ 'A1','A2','B1','B2','C1','C2','C3' ]

So I have some big categories `'A', 'B', 'C'` and for each, I have two (or more) labels.

Now, what I want to do is to compute the cross entropy loss considering confusions only in the labels within a big category. So if a training sample has `y_true=A1` I will apply a mask `mask_vector=[1,1,0,0,0,0,0]`, to keep the probabilities for `A1` and `A2`. After this I should re-normalize so `P(A1) + P(A2) = 1` and then compute the cross entropy loss.

So I have a confusion cost matrix that looks like the following

        --  A1   A2   B1   B2   C1  C2  C3
        A1   1    1    0    0    0   0   0
        A2   1    1    0    0    0   0   0
        B1   0    0    1    1    0   0   0
    M = B2   0    0    1    1    0   0   0
        C1   0    0    0    0    1   1   1
        C2   0    0    0    0    1   1   1
        C3   0    0    0    0    1   1   1

so if i write the equations `y_true = [1,0,0,0,0,0,0]` then

                         //index the correct column from M     //mask proba
    cross_entropy(y_true, matmul( y_true,M )                   * y_pred     )

----------

There are two constraints that are not explicit above

 1. I have to handle the batch size, which has been problematic
 2. The model is Seq-to-Seq, so the X has 3 axis `(batch_size,sequence_length,nb_labels)`  
 

----------

So this is what I implemented. I have a function that computes the M matrix once time and then I return my custom loss with this M encoded within.

      label2idx = { 'A1':0,'A2':1,'B1':2,'B2':3,'C1':4,'C2':5,'C3':6 }

      def custom_rec_gan_loss( label2idx ):
        nb_labels   = len( label2idx.keys() )
        cost_matrix = np.zeros( (1,nb_labels,nb_labels), dtype=np.float32 )
        for label1, i in label2idx.items():
            for label2, j in label2idx.items():
                if( label1[0] == label2[0] ):
                    cost_matrix[0][i][j] = 1
    
        M = tf.convert_to_tensor(cost_matrix, dtype=tf.float32)
    
        def my_loss( y_true, y_pred ):
            return K.categorical_crossentropy(y_true, K.tf.matmul(y_true, tf.tile( M ,[ tf.shape(y_true)[0] ,1,1]) ) *y_pred )
    
    
        return my_loss

So I detail the important line of my_loss function:

    #compute batch_size 
    batch_size = tf.shape(y_true)[0]

    #stack the M matrix as many times as the batch size
    #at this point I dont know what happens with the sequence_length parameter
    stacked_M = tf.tile( M ,[batch_size, 1, 1] )

    #multiply the stacked_M with the y_true labels to obtain the masks
    my_batch_masks = K.tf.matmul(y_true, stacked_M ) 

    #apply the mask using elemnt-wise multiplication
    batch_masked_probas = my_batch_masks*y_pred

    #compute cross entropy, tf documentation says this handles normalization
    K.categorical_crossentropy(y_true, batch_masked_probas )

So I get to train my sequence tagger using this loss but when I use it for prediction I get NANs as output probabilities.
I don't know what can be causing this issue, I guess it may be an error handling the sequence length or during the normalization of the output probabilities but I have no clue and I'm having a hard time to debug this.

Any help is appreciated,
Thank you very much for your attention,
Gabriel
